plugins/obs-filters/crosshatching-filter.c

```diff
@@ -0,0 +1,254 @@
+
+#define OMT                             obs_module_text
+#define TEXT_OFFSET                     OMT("Offset")
+#define TEXT_THRESHOLD_1                OMT("Threshold.1")
+#define TEXT_THRESHOLD_2                OMT("Threshold.2")
+#define TEXT_THRESHOLD_3                OMT("Threshold.3")
+#define TEXT_THRESHOLD_4                OMT("Threshold.4")
+
+
+struct crosshatching_filter_data {
+	obs_source_t                   *context;
+
+	gs_effect_t                    *effect;
+
+	gs_eparam_t                    *offset_param;
+	gs_eparam_t                    *threshold_1_param;
+	gs_eparam_t                    *threshold_2_param;
+	gs_eparam_t                    *threshold_3_param;
+	gs_eparam_t                    *threshold_4_param;
+
+	uint32_t                        offset;
+	float                           threshold_1;
+	float                           threshold_2;
+	float                           threshold_3;
+	float                           threshold_4;
+};
+
+
+/* As the functions' namesake, this provides the user facing name
+ * of your Filter. */
+static const char *crosshatching_filter_name(void *unused)
+{
+	UNUSED_PARAMETER(unused);
+	return obs_module_text("Crosshatching");
+}
+
+/* This function is called (see bottom of this file for more details
+ * whenever the OBS filter interface changes. So when the user is messing
+ * with a slider this function is called to update the internal settings
+ * in OBS, and hence the settings being passed to the CPU/GPU. */
+static void crosshatching_filter_update(void *data, obs_data_t *settings)
+{
+	struct crosshatching_filter_data *filter = data;
+	filter->offset = (uint32_t)obs_data_get_int(settings,
+			SETTING_OFFSET);
+	filter->threshold_1 = (float)obs_data_get_double(settings,
+			SETTING_THRESHOLD_1);
+	filter->threshold_2 = (float)obs_data_get_double(settings,
+			SETTING_THRESHOLD_2);
+	filter->threshold_3 = (float)obs_data_get_double(settings,
+			SETTING_THRESHOLD_3);
+	filter->threshold_4 = (float)obs_data_get_double(settings,
+			SETTING_THRESHOLD_4);
+}
+
+/* Since this is C we have to be careful when destroying/removing items from
+ * OBS. Jim has added several useful functions to help keep memory leaks to
+ * a minimum, and handle the destruction and construction of these filters. */
+static void crosshatching_filter_destroy(void *data)
+{
+	struct crosshatching_filter_data *filter = data;
+
+	if (filter->effect) {
+		obs_enter_graphics();
+		gs_effect_destroy(filter->effect);
+		obs_leave_graphics();
+	}
+
+	bfree(data);
+}
+
+/*
+ * When you apply a filter OBS creates it, and adds it to the source. OBS also
+ * starts rendering it immediately. This function doesn't just 'create' the
+ * filter, it also calls the render function (farther below) that contains the
+ * actual rendering code.
+ */
+static void *crosshatching_filter_create(obs_data_t *settings, obs_source_t *context)
+{
+	/*
+	 * Because of limitations of pre-c99 compilers, you can't create an
```

Comment:
> pre-c99 compilers

In at least one other file you capitalize this as "pre-C99", which is probably how it should be written.